reworks js in the browser day (#693)
Ada-Developers-Academy/textbook-curriculum: 10-JavaScript/js-in-the-browser.md

- A lack of standards, process of standardization, or regulation around web technologies

Events in the 90s led to the major browser, Netscape, and thus its competitors to adopt support for JavaScript. Meaning, **since early in the history of web, browsers became environments that could read and execute JavaScript code.**

The development and evolution of browsers, browser features, and JavaScript's language and utilities have been intertwined. Browsers, by necessity, support and run JavaScript, and evolve to do so in nicer and more efficient ways.

This should feel special to us because browsers do not run other languages like this "out of the box" (without special tools, libraries, or circumstances). Our Rails web apps never asked _the browser_ to run any Ruby code; it was always our servers (or our machines/computers that acted as web servers) that needed to execute Ruby. Our servers simply **gave responses with website/HTML data,** which our browsers are very good at doing.

JavaScript is now:
- **The** programming language for modern web
- Standardized and open
- An evolved language with many more features

### Exploring Chrome Dev Tools

> Let's prove to ourselves that JavaScript can run in the browser by executing JavaScript in the browser. We can do this with Chrome Dev Tools's console. It doesn't run Ruby code.

Reviewer: It feels a little weird to say it specifically doesn't run Ruby code.  It makes it sound like it maybe supports multiple languages and Ruby just isn't one.

```suggestion
Let's prove to ourselves that JavaScript can run in the browser by executing JavaScript in the browser. We can do this with Chrome Dev Tools's console (it only runs JavaScript code).
```

Go to **any** webpage, such as Github.com. Note that we are not asking "how do we install Javascript?"

> Open Chrome Dev Tools. Note the following tabs on the top (written 2019):

Reviewer: ```suggestion
Open Chrome Dev Tools. Note the following tabs on the top (as of 2019):
```

| Memory | details on memory usage and allocation, and tracking memory leaks
| Application | details on things like local storage, cookies, cache, etc.

... and more!

Navigate to the console, either through the tab, or through other means (such as the Elements tab).

Within it, play around with it as a REPL and execute JavaScript in it:

- `2 + 3`
- `const fruit = 'apples'`
- `let color = 'red'`
- <code>\`${color} ${fruit}`</code>
- `color = 'pink'`
- `fruit = 'dragonfruit'`
- `const helloWorld = () => 'foo'`
- `helloWorld()`

Now try to execute Ruby in it:

- `[1, 2, 3].sample`
- `def hello_world`

And we should see that Ruby can't run in the browser.

## Browsers Listen for User Events

Another way to think about building websites is "when this event happens, then this should should happen."

### Event-Driven Programming

- When does a specific element trigger a specific event?
  - What is the element? What is the event name?
- How is the event handled? We need a function that is listening for a specific event name on a specific element.

## HTML and the DOM

Browsers do a thing: They parse HTML documents as DOM trees

> https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model

Reviewer: ```suggestion
[MDN: Document Object Model (DOM)](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model)
```

> https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction

Reviewer: ```suggestion
[MDN's Introduction to the DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction)
```

- A lack of standards, process of standardization, or regulation around web technologies

Events in the 90s led to the major browser, Netscape, and thus its competitors to adopt support for JavaScript. Meaning, **since early in the history of web, browsers became environments that could read and execute JavaScript code.**

The development and evolution of browsers, browser features, and JavaScript's language and utilities have been intertwined. Browsers, by necessity, support and run JavaScript, and evolve to do so in nicer and more efficient ways.

This should feel special to us because browsers do not run other languages like this "out of the box" (without special tools, libraries, or circumstances). Our Rails web apps never asked _the browser_ to run any Ruby code; it was always our servers (or our machines/computers that acted as web servers) that needed to execute Ruby. Our servers simply **gave responses with website/HTML data,** which our browsers are very good at doing.

JavaScript is now:
- **The** programming language for modern web
- Standardized and open

> - An evolved language with many more features

Reviewer: Insert an image of pokemon evolutions here?

Nodes in the DOM have properties that match the attributes set on them in the HTML (e.g. an `img` node might have `src` and `id` properties). DOM nodes also have children, which are all of the other nodes nested within them (e.g. a `div` node might have a `table` child node).

Run `document` in the console. What do you get?

## Browser-Supported Ways to Manipulate the DOM with JS

Logging to the console is alright, but the true power of JavaScript is that it can dynamically change the contents of the webpage.

Browsers have already defined a way (an interface) to represent, access, and modify the contents of a website. This interface is called **the DOM**. When we use this interface, we can dynamically change the page using JS.

Until now our web applications have rendered _static_ HTML which defines a webpage. They have then sent it to the browser which in turn "builds the DOM" from that HTML.

> When we use JavaScript to manipulate the DOM, we do not change the original HTML that was sent to the browser, instead we change the browser's internal representation of that same webpage. An important consequence of this distinction is that all of our changes disappear as soon as the browser forgets about that webpage (e.g. if we close the tab or the whole browser).

Reviewer: I think it's important to call out refreshing as something that will lose their state.

```suggestion
When we use JavaScript to manipulate the DOM, we do not change the original HTML that was sent to the browser, instead we change the browser's internal representation of that same webpage. An important consequence of this distinction is that all of our changes disappear as soon as the browser forgets about that webpage (e.g. if we close the tab/browser or refresh the page).
```

> Web API: https://developer.mozilla.org/en-US/docs/Web/API

Reviewer: ```suggestion
[MDN: Web API](https://developer.mozilla.org/en-US/docs/Web/API)
```

> Document API: https://developer.mozilla.org/en-US/docs/Web/API/Document

Reviewer: ```suggestion
[MDN: Document API](https://developer.mozilla.org/en-US/docs/Web/API/Document)
```

### Default Events the Browser Listens for

> https://developer.mozilla.org/en-US/docs/Web/Events

Reviewer: ```suggestion
[MDN: Event Reference](https://developer.mozilla.org/en-US/docs/Web/Events)
```
